feat: add support for 2.6.0 and OPA authorizer (#506)
stackabletech/hbase-operator: docs/modules/hbase/pages/usage-guide/security.adoc

We have an https://github.com/stackabletech/hbase-operator/blob/main/tests/templates/kuttl/kerberos/41-access-hbase.yaml.j2[integration test] for this exact purpose, where you can see how to connect and get a valid keytab.

== Authorization

> NOTE: This is an exprimental feature available from HBase version 2.6.0 upwards.

Reviewer: ```suggestion
NOTE: This is an experimental feature available from HBase version 2.6.0 upwards.
```

We have an https://github.com/stackabletech/hbase-operator/blob/main/tests/templates/kuttl/kerberos/41-access-hbase.yaml.j2[integration test] for this exact purpose, where you can see how to connect and get a valid keytab.

== Authorization

> Authorization using the Open Policy Agent (OPA) is supported.

Reviewer: I think we should have an INFO box mentioning that this is only supported as of 2.6 which is experimental right now.

Author: fixed https://github.com/stackabletech/hbase-operator/pull/506/commits/ac80b16b594c641bff70b2d164b6cb99c483773e

> A custom https://github.com/stackabletech/hbase-opa-authorizer[access controller CoProcessor] delegates authorization decisions to OPA.

Reviewer: ```suggestion
A custom https://github.com/stackabletech/hbase-opa-authorizer[access controller coprocessor] delegates authorization decisions to OPA.
```

For more information on the OPA authorization mechanism and how to use it within the Stackable Platform, see the xref:concepts:opa.adoc[OPA concepts] documentation.

IMPORTANT: Authentication with Kerberos is required when using authorization.

> Authorization configurations is split in two parts:

Reviewer: ```suggestion
Authorization configuration is split in two parts:
```

The operator automatically configures the HBase coprocessor with the following properties:

* `hbase.coprocessor.master.classes`, `hbase.coprocessor.region.classes` and `hbase.coprocessor.regionserver.classes` are set to `tech.stackable.hbase.OpenPolicyAgentAccessController`
* `hbase.security.authorization.opa.policy.url` is read from the discovery ConfigMap of the OPA cluster
* `hbase.security.authorization.opa.policy.dryrun` : `false`
* `hbase.security.authorization.opa.policy.cache.active` : `true`
* `hbase.security.authorization.opa.policy.cache.seconds` : `300` (5 minutes)
* `hbase.security.authorization.opa.policy.cache.size` : `1000`

You can override these properties for each HBase role using the `<role>.configOverrides.hbase-site.xml` property.

=== Rego rules

Currently, rego rules are deployed as ConfigMaps that are automatically discovered by the OPA cluster.
This guide assumes your OPA cluster is called `opa`.

[source,rego]
----
include::example$usage-guide/hbase-regorules.yaml[]
----

This rego rule is intended for demonstration purposes and allows every operation.
For a production setup you will probably need to have something much more granular.

> We provide a more representative rego rule in our integration tests and in the aforementioned co-processor repository.

Reviewer: ```suggestion
We provide a more representative rego rule in our integration tests and in the aforementioned coprocessor repository.
```

WARNING: This implementation takes an approach to HBase authorization that is fundamentally different from the default one (which relies on HDFS group mappings) and departs significantly from traditional Hadoop patterns and POSIX-style permissions.

> The current rego rules ignore file ownership and permissions, and ACLs are persisted in neither ZooKeeper nor internal HBase tables.

Reviewer: ```suggestion
The current rego rules ignore file ownership and permissions, and ACLs are persisted neither in ZooKeeper nor internal HBase tables.
```

Keeping this state in HDFS/HBase clashes with the infrastructure-as-code approach (IaC).

> Instead, HBase will send a request detailing who (e.g. `alice/test-hbase-permissions.default.svc.cluster.local@CLUSTER.LOCAL`) is trying to execute what type of action (e.g. `READ`, `WRITE`, `CREATE` or `ADMIN`) on what namespace or table (e.g. `developers:`, `developers/table1`).

Reviewer: ```suggestion
Instead, HBase will send a request detailing who (e.g. `alice/test-hbase-permissions.default.svc.cluster.local@CLUSTER.LOCAL`) is trying to execute what type of action (e.g. `READ`, `WRITE`, `CREATE` or `ADMIN`) on what namespace or table (e.g. `developers:`, `developers/table1`) to OPA.
```

> OPA then makes a decision if this action is allowed or not.

Reviewer: ```suggestion
OPA then makes a decision whether this action is allowed or not.
```

> Instead of using the HBase shell to grant or revoke rights to users, you can go to your IaC Git repository and add a rego rule entry specifying what a specific user is allowed to do to which specific resources.

Reviewer: This makes assumptions.
I think we should write something along the lines of 

```suggestion
Instead of using the HBase shell to grant or revoke rights to users, you can create rules for OPA using the Rego language to define what a specific user is allowed or not allowed to do to.
```

}

# Identity checks based on e.g.
# - explicit matches on the fully-qualified userName
# - regex matches
# - the group membership (simple- or regex-matches on fully-qualified userName)
matches_identity(identity) if {
    ...
}

# Resource checks on e.g.
# - wildcard access (for e.g. admins)
# - explicit namespace- or table-mentions
matches_resource(file, resource) if {
    ...
}

# Check the operation and its implicit action against an ACL
action_sufficient_for_operation(action, operation) if {
    action_hierarchy[action][_] == action_for_operation[operation]
}

action_hierarchy := {
    "full": ["full", "rw", "ro"],
    "rw": ["rw", "ro"],
    "ro": ["ro"],
}

action_for_operation := {
    "ADMIN": "full",
    "CREATE": "full",
    "WRITE": "rw",
    "READ": "ro",
}

# Maps users to groups
groups_for_user := {
    "fully-qualified userName": ["ACL group"],
    ...
}

acls := [
    {
        # Identity: either "group:groupName" or "user:userName"
        "identity": "group:admins",
        "action": "full",
        "resource": ":",
    },
    {
        "identity": "group:developers",
        "action": "full",
        "resource": "developers:",
    },
    ...
]
----

> There is complete file shown below, together with a list of test cases.

Reviewer: ```suggestion
There is a complete file shown below, together with a list of test cases.
```

[source,rego]
----
include::example$rego/hbase.rego[]
----
</details>

<details>
<summary>Rego tests</summary>
[source,rego]
----
include::example$rego/hbase_test.rego[]
----
</details>

If the https://www.openpolicyagent.org/docs/latest/cli/[Opa CLI tool] has been installed, the tests can be executed by calling the following from the root folder:

[shell]
----
opa test ./docs/modules/hbase/examples/rego/ -bv

e.g.
17:44 $ opa test ./docs/modules/hbase/examples/rego/ -bv
docs/modules/hbase/examples/rego/hbase_test.rego:
data.hbase.test_permission_admin: PASS (2.055153ms)
data.hbase.test_namespace_admin: PASS (528.212µs)
data.hbase.test_permission_developers: PASS (896.004µs)
data.hbase.test_permission_alice: PASS (1.56663ms)
data.hbase.test_no_permission_bob: PASS (787.445µs)
data.hbase.test_permission_bob1: PASS (1.485691ms)
data.hbase.test_permission_bob2: PASS (1.459236ms)
data.hbase.test_permission_hbase: PASS (1.524541ms)
data.hbase.test_permission_testuser: PASS (1.036205ms)
data.hbase.test_permission_readonlyuser1: PASS (780.05µs)
data.hbase.test_permission_readonlyuser2: PASS (1.37649ms)
--------------------------------------------------------------------------------

> PASS: 11/11

Reviewer: Does this work? The syntax highlighting in the Github UI suggests that the previous dashes already ended the Asciidoc block.
Have you tried rendering this?

Author: fixed https://github.com/stackabletech/hbase-operator/pull/506/commits/ac80b16b594c641bff70b2d164b6cb99c483773e
